pkgtools/pkg_filecheck/files/pkg_filecheck.1
.\"	$NetBSD: pkg_filecheck.1,v 1.1.1.1 2003/07/13 16:15:34 wiz Exp $
.\"
.\" Copyright (c) 2003 Thomas Klausner
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR, THOMAS KLAUSNER,
.\" ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
.\" TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
.\" BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd July 13, 2003
.Dt PKG_FILECHECK 1
.Os
.Sh NAME
.Nm pkg_filecheck
.Nd compare file system contents with package database
.Sh SYNOPSIS
.Nm
.Op Fl h
.Op Fl c Ar config-file
.Sh DESCRIPTION
.Nm
reads the package tools database and and compares it with the actual
contents of the directory hierarchy starting at
.Pa @PREFIX@ ,
reporting any files that are on the file system, but not in the
database.
The database used and the hierarchy to be checked can be changed
in the config file.
.Pp
Supported options:
.Bl -tag -width XcXconfigXfileX -offset indent
.It Fl c Ar config-file
Use
.Ar config-file
as config file instead of the default
.Pa @PKG_SYSCONFDIR@/pkg_filecheck.conf .
.It Fl h
Display a very short help message.
.El
.Sh SEE ALSO
.Xr builddb 1
.Sh BUGS
.Nm
does not report files that are in the database, but not in the file
system.
You can use
.Dq Ic pkg_admin Ar check
for that.


doc/NEWS
# $NetBSD: NEWS,v 1.33 2025/12/03 21:57:26 leot Exp $

PKGSRC NEWS
===========

  This file contains information about important infrastructural changes in
  pkgsrc. The intended readers are the pkgsrc developers as well as anyone
  tracking pkgsrc-current.

m68k softfloat (m68ksf) support [nat 2025-10-22]

        Support for m68k softfloat was added to pkgsrc.

Default versions bumped [wiz 2025-10-06]

	PYTHON_VERSION_DEFAULT is now 313.

Default MySQL implementation changed to MariaDB 11.4 [nia 2024-12-25]

	The default version of MySQL as set by the pkgsrc infrastructure
	was changed from MariaDB 10.11 to MariaDB 11.4.

Default versions bumped [wiz 2024-08-12]

	PYTHON_VERSION_DEFAULT is now 312.

Default versions bumped [taca 2024-02-10]

	RUBY_VERSION_DEFAULT is now 32.

Default MySQL implementation changed to MariaDB 10.11 [nia 2024-01-23]

	The default version of MySQL as set by the pkgsrc infrastructure
	was changed from MariaDB 10.6 to MariaDB 10.11.

Default versions bumped [wiz 2023-08-14]

	PYTHON_VERSION_DEFAULT is now 311.

Support for USE_CC_FEATURES and USE_CXX_FEAUTRES [nia 2023-07-18]

	USE_CC_FEATURES and USE_CXX_FEATURES can now be used
	to specify a minimum level of C and C++ support required
	to compile a package, including specific libstdc++ features.

	ALLOW_NEWER_COMPILER can be set to no in mk.conf to skip
	building packages that require a newer compiler than the system
	provides.

Default versions bumped [nia 2022-06-30]

	PYTHON_VERSION_DEFAULT is now 310.
	PGSQL_VERSION_DEFAULT is now 14.
	LUA_VERSION_DEFAULT is now 54.

Dropped support for compiling glib2-using packages without introspection
support [wiz 2022-06-07]

PKGSRC_USE_MKTOOLS support added [jperkin 2022-06-07]

        Added C-based tools to speed up various parts of the
        mk infrastructure via PKGSRC_USE_MKTOOLS=yes in mk.conf.

        For packages with large numbers of dependencies, you should see
        anywhere up to around 10x speed-up for the 'wrapper' phase, and
        perhaps more importantly a huge reduction in system time usage.

Linux infrastructure adapted to non-standard distributions [nia 2022-05-01]

        The Linux pkgsrc infrastructure was adapted to support Linux
        distributions where binaries and shared objects are not
        stored in a normal Unix-like file system layout.  It's hoped
        that this will be useful for pkgsrc development on platforms
        like NixOS and Gobo Linux.

UnixWare support [nia 2021-12-26]

        Support for SCO/Xinuos UnixWare was improved to the point where
        it is possible to complete a bootstrap on this platform.

SHA1 distfile hashes deprecated [nia 2021-10-07]

        Distfile hash algorithms have been changed, keeping SHA512,
        adding BLAKE2s, and removing SHA1 and RMD160.  New packages
        and updates will only generate SHA512 and BLAKE2s, and the
        stored hashes have been adjusted.

Default Python version changed to 3.9 [nia 2021-10-01]

Increased enabled-by-default hardening options [nia 2021-09-27]

	PKGSRC_USE_MKPIE was enabled on some architectures.
	Packages that provide static libraries to other packages
	should be rebuilt with MKPIE enabled.
	Some packages may experience runtime errors on platforms
	like i386 - usually this is due to unsafe assembly code,
	which should be disabled when MKPIE is enabled.

	PKGSRC_USE_SSP was bumped from "yes" to "strong".  This
	is not expected to have a noticable effect.

	More information:
	https://www.NetBSD.org/docs/pkgsrc/hardening.html
	(or see "Security hardening" section of doc/pkgsrc.txt)

Default MySQL implementation changed to MariaDB 10.6 [nia 2021-09-27]

Default PostgreSQL changed to 13 [nia 2021-06-28]

Default MySQL implementation changed to MariaDB [nia 2021-06-24]

	The default version of MySQL as set by the pkgsrc infrastructure
	was changed from MySQL 5.7 to MariaDB 10.5.

	This is configurable by setting MYSQL_VERSION_DEFAULT in mk.conf.

	PHP packages are no longer tied to a specific MySQL version
	due to the native driver being enabled by default.

Python 3.8 default [nia 2020-12-04]
	The default Python version was switched to 3.8.

Default database location changed for NetBSD [wiz 2020-12-02]
	The database of installed packages now defaults to ${PREFIX}/pkgdb
	instead of /var/db/pkg on NetBSD. This matches the default on
	other operating systems and cases where a non-default bootstrap
	is used.

	To deal with this change, see this blog post:

	https://pkgsrc.org/news/pkgdb-change/

	Since this change affects the packaging tools only, there is no change
	to binary packages and they do not need to be rebuilt.

lua 5.3 default [nia 2020-03-30]
	The default Lua version was switched to 5.3.
	Since many packages specify a preferred Lua version, this is mainly
	expected to affect users of Lua modules who were using the default
	settings.

gfortran default [nia 2020-03-30]
	gfortran is now the default Fortran compiler instead of g95.
	gfortran is expected to support more pkgsrc platforms and be
	better at compiling newer Fortran code.

python 3.7 default [maya 2019-04-24]
	As Python 2.x versions are nearing end of life, we have switched
	the default to a newer version.
	Python 2.7 packages are still available, and most python packages
	are built for all available python versions.

Lua multiversion support [alnsn 2014-05-03]
	Lua packages lang/lua51 and lang/lua52 have been modified
	to coexist without conflicts, infrastructure for supporting
	multiple versions of Lua has been put in place and all
	dependent packages have been adapted to the new infrastructure.

SMF support is introduced [jperkin 2014-03-11]

Unprivileged staged installation is default [asau 2012-08-14]
	pkgsrc has changed to use unprivileged staged installation by default.
	This makes it possible to prevent some installation mistakes.
	It also makes detection of package conflicts easier.
	Previously you had to set PKG_DESTDIR_SUPPORT to "user-destdir"

pkgtools/pkg_filecheck/files/pkg_filecheck.c
/*
  $NetBSD: pkg_filecheck.c,v 1.5 2007/02/16 22:57:09 tonio Exp $

  pkg_filecheck.c -- check for files not owned by any package
  Copyright (C) 2001 Dieter Baron

  Redistribution and use in source and binary forms, with or without
  modification, are permitted provided that the following conditions
  are met:
  1. Redistributions of source code must retain the above copyright
     notice, this list of conditions and the following disclaimer.
  2. Redistributions in binary form must reproduce the above copyright
     notice, this list of conditions and the following disclaimer in
     the documentation and/or other materials provided with the
     distribution.
  3. The name of the author may not be used to endorse or promote
     products derived from this software without specific prior
     written permission.
 
  THIS SOFTWARE IS PROVIDED BY DIETER BARON ``AS IS'' AND ANY EXPRESS
  OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
  WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
  ARE DISCLAIMED.  IN NO EVENT SHALL DIETER BARON BE LIABLE FOR ANY
  DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
  GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
  IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
  OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
  IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

#if defined(USE_LIBNBCOMPAT)
#include <nbcompat.h>
#endif

#include <sys/types.h>
#include <sys/stat.h>
#include <db_185.h>
#include <errno.h>
#include <fcntl.h>
#if !defined(USE_LIBNBCOMPAT) || defined(HAVE_FTS_H)
#include <fts.h>
#endif
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

struct array {
    void **p;
    int len, alen;
};

char *prg;

static int check(char **dir, DB **db, char **ignore);
static int fts_sort(const FTSENT **e1, const FTSENT **e2);
static void push(struct array *a, void *el);
static int read_conf(char *fname, char ***dirp, DB ***dbp, char ***ignorep);



int
main(int argc, char *argv[])
{
    DB **db;
    char **dir, **ignore;

    char *cfgfile;
    int c, i, err;

    prg = argv[0];
    opterr = 0;

    cfgfile = "@PKG_SYSCONFDIR@/pkg_filecheck.conf";

    while ((c=getopt(argc, argv, "c:h")) != -1) {
	switch (c) {
	case 'c':
	    cfgfile = optarg;
	    break;

	case 'h':
	    printf("usage: %s [-h] [-c config-file]\n", prg);
	    exit(0);

	default:
	    fprintf(stderr, "usage: %s [-h] [-c config-file]\n", prg);
	    exit(1);
	}
    }

    if (read_conf(cfgfile, &dir, &db, &ignore) < 0)
	exit(1);

    err = check(dir, db, ignore);

    for (i=0; db[i]; i++)
	db[i]->close(db[i]);
    free(dir);
    free(db);
    free(ignore);

    exit(err ? 1 : 0);
}



static int
check(char **dir, DB **db, char **ignore)
{
    FTS *fts;
    FTSENT *ent;
    DBT key, val;
    int i, found;

    if ((fts=fts_open(dir, FTS_PHYSICAL, fts_sort)) == NULL) {
	fprintf(stderr, "%s: cannot walk direcotry hierarchy: %s",
		prg, strerror(errno));
	return -1;
    }

    while ((ent=fts_read(fts))) {
	switch (ent->fts_info) {
	case FTS_F:
	    key.data = ent->fts_path;
	    key.size = ent->fts_pathlen+1;
	    
	    found = 0;
	    for (i=0; db[i]; i++)
		if (db[i]->get(db[i], &key, &val, 0) == 0) {
		    found = 1;
		    break;
		}

	    if (!found)
		printf("%s\n", ent->fts_path);
	    break;

	case FTS_D:
	    if (ignore)
		for (i=0; ignore[i]; i++)
		    if (strcmp(ignore[i], ent->fts_path) == 0) {
			fts_set(fts, ent, FTS_SKIP);
			break;
		    }
	    break;

	case FTS_DNR:
	    fprintf(stderr, "%s: cannot read directory `%s', skipped: %s\n",
		    prg, ent->fts_path, strerror(ent->fts_errno));
	    break;

	case FTS_NS:
	    fprintf(stderr, "%s: cannot stat file `%s', skipped: %s\n",
		    prg, ent->fts_path, strerror(ent->fts_errno));
	    break;

	case FTS_ERR:
	    fprintf(stderr, "%s: directory traversal error: %s\n",
		    prg, strerror(ent->fts_errno));
	    break;

	case FTS_DC:
	case FTS_DP:
	case FTS_SL:
	case FTS_SLNONE:
	case FTS_DEFAULT:
	    /* ignore */
	    break;

	default:
	    fprintf(stderr, "%s: unknown FTSENT type %d ignored\n",
		    prg, ent->fts_info);
	    break;
	}
    }

    fts_close(fts);
    return 0;
}



static int
fts_sort(const FTSENT **e1, const FTSENT **e2)
{
    return strcmp((*e1)->fts_name, (*e2)->fts_name);
}



static int
read_conf(char *fname, char ***dirp, DB ***dbp, char ***ignorep)
{
    struct array dir, db, ignore;
    FILE *f;
    DB *cdb;
    char b[8192], *curd, *dbname, *cmd, *p;
    int lineno, err;

    dir.p = db.p = ignore.p = NULL;
    dir.len = db.len = ignore.len = 0;
    dir.alen = db.alen = ignore.alen = 0;

    if ((f=fopen(fname, "r")) == NULL) {
	fprintf(stderr, "%s: cannot open config file `%s': %s\n",
		prg, fname, strerror(errno));
	return -1;
    }

    curd = NULL;

    lineno = 0;
    while (fgets(b, 8192, f)) {
	lineno++;
	if (b[0] == '#')
	    continue;

	cmd = b+strspn(b, " \t");
	p = cmd+strcspn(cmd, " \t\n");
	*(p++) = '\0';
	p += strspn(p, " \t");
	if (*p == '\n' || *p == '\0')
	    continue;

	if (p[strlen(p)-1] == '\n')
	    p[strlen(p)-1] = '\0';

	if (strcmp(cmd, "dir") == 0) {
	    free(curd);
	    curd = strdup(p);
	}
	else if (strcmp(cmd, "db") == 0) {
	    if (curd) {
		dbname = malloc(strlen(curd) + strlen(p) + 2);
		sprintf(dbname, "%s/%s", curd, p);
	    }
	    else
		dbname = p;
	    if ((cdb=dbopen(dbname, O_RDONLY, 0, DB_BTREE, NULL)) == NULL) {
		fprintf(stderr, "%s:%s:%d: cannot open data base `%s': %s\n",
			prg, fname, lineno, dbname, strerror(errno));
		free(dir.p);
		free(db.p);
		free(ignore.p);
		return -1;
	    }
	    else
		push(&db, cdb);

	    if (curd)
		free(dbname);
	}
	else if (strcmp(cmd, "check") == 0)
	    push(&dir, strdup(p));
	else if (strcmp(cmd, "ignore") == 0)
	    push(&ignore, strdup(p));
	else {
	    fprintf(stderr, "%s:%s:%d: unrecognized command `%s'\n",
		    prg, fname, lineno, cmd);
	}
    }

    fclose(f);

    err = 0;

    if (dir.p == NULL) {
	fprintf(stderr, "%s: no directories to check\n",
		prg);
	err = 1;
    }
    if (db.p == NULL) {
	fprintf(stderr, "%s: no file data bases given\n",
		prg);
	err = 1;
    }

    if (err) {
	free(dir.p);
	free(db.p);
	free(ignore.p);
	return -1;
    }

    *dirp = (char **)dir.p;
    *dbp = (DB **)db.p;
    *ignorep = (char **)ignore.p;

    return 0;
}



static void
push(struct array *a, void *el)
{
    int alen;
    void *p;

    if (a->len + 2 > a->alen) {
	alen = (a->alen == 0) ? 8
	     : (a->alen < 1024) ? (a->alen * 2)
	     : (a->alen + 1024);

	if ((p=realloc(a->p, alen*sizeof(void *))) == NULL) {
	    fprintf(stderr, "%s: malloc failure\n", prg);
	    exit(1);
	}
	
	a->p = p;
	a->alen = alen;
    }

    a->p[a->len++] = el;
    a->p[a->len] = NULL;
}
